File: frontend/src/types/question.ts
export type QuestionType =
  | 'MCQ'
  | 'TRUE_FALSE'
  | 'FILL_BLANK'
  | 'SHORT_ANSWER'
  | 'DESCRIPTIVE'
  | 'PROBLEM_SOLVING'
  | 'NUMERICAL'
  | 'PROGRAMMING'
  | 'CASE_STUDY';

export type DifficultyLevel = 'EASY' | 'MEDIUM' | 'HARD';

export type BloomLevel =
  | 'REMEMBER'
  | 'UNDERSTAND'
  | 'APPLY'
  | 'ANALYZE'
  | 'EVALUATE'
  | 'CREATE';

export type QuestionStatus =
  | 'DRAFT'
  | 'AI_GENERATED'
  | 'UNDER_REVIEW'
  | 'APPROVED'
  | 'ACTIVE'
  | 'REJECTED'
  | 'ARCHIVED';

export interface MCQOptions {
  options: string[];
  correct_option: string;
  explanation?: string;
}

export interface ValidationData {
  is_valid: boolean;
  alignment_score: number;
  difficulty_score: number;
  bloom_alignment_score: number;
  duplicate_score: number;
  completeness_score: number;
  warnings: string[];
  similar_question_id?: string;
}

export interface QuestionVersion {
  id: string;
  question_id: string;
  version_number: number;
  question_text: string;
  marks: number;
  difficulty: DifficultyLevel;
  bloom_level: BloomLevel;
  expected_answer?: string;
  options?: MCQOptions;
  keywords?: string[];
  concepts?: string[];
  changed_by?: string;
  change_reason?: string;
  created_at: string;
}

export interface Question {
  id: string;
  subject_id: string;
  unit_id?: string;
  topic_id?: string;
  learning_outcome_id?: string;
  generation_run_id?: string;
  question_text: string;
  question_type: QuestionType;
  marks: number;
  difficulty: DifficultyLevel;
  bloom_level: BloomLevel;
  expected_answer?: string;
  options?: MCQOptions;
  keywords?: string[];
  concepts?: string[];
  source?: string;
  status: QuestionStatus;
  version: number;
  validation_data?: ValidationData;
  created_by?: string;
  versions?: QuestionVersion[];
  created_at: string;
  updated_at: string;
}

export interface QuestionListResponse {
  total: number;
  items: Question[];
  page: number;
  page_size: number;
}

export interface QuestionCreateRequest {
  subject_id: string;
  unit_id?: string;
  topic_id?: string;
  learning_outcome_id?: string;
  question_text: string;
  question_type: QuestionType;
  marks: number;
  difficulty: DifficultyLevel;
  bloom_level: BloomLevel;
  expected_answer?: string;
  options?: MCQOptions;
  keywords?: string[];
  concepts?: string[];
  status?: QuestionStatus;
  source?: string;
}

export interface QuestionUpdateRequest {
  subject_id?: string;
  unit_id?: string;
  topic_id?: string;
  learning_outcome_id?: string;
  question_text?: string;
  question_type?: QuestionType;
  marks?: number;
  difficulty?: DifficultyLevel;
  bloom_level?: BloomLevel;
  expected_answer?: string;
  options?: MCQOptions;
  keywords?: string[];
  concepts?: string[];
  status?: QuestionStatus;
  change_reason?: string;
}

export interface QuestionGenerationRequest {
  subject_id: string;
  unit_id?: string;
  topic_id?: string;
  learning_outcome_id?: string;
  question_type: QuestionType;
  difficulty: DifficultyLevel;
  bloom_level: BloomLevel;
  marks: number;
  count: number;
  language?: string;
}

export interface QuestionGenerationRun {
  id: string;
  subject_id: string;
  unit_id?: string;
  topic_id?: string;
  learning_outcome_id?: string;
  created_by: string;
  provider: string;
  model: string;
  prompt_version: string;
  status: string;
  requested_count: number;
  generated_count: number;
  accepted_count: number;
  rejected_count: number;
  error_message?: string;
  questions?: Question[];
  created_at: string;
  updated_at: string;
}

export interface QuestionBankItem {
  id: string;
  question_bank_id: string;
  question_id: string;
  added_by?: string;
  question?: Question;
  created_at: string;
}

export interface QuestionBank {
  id: string;
  subject_id: string;
  name: string;
  description?: string;
  created_by?: string;
  items?: QuestionBankItem[];
  created_at: string;
  updated_at: string;
}

export interface QuestionBankCreateRequest {
  subject_id: string;
  name: string;
  description?: string;
}

File: backend/app/schemas/question.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, ConfigDict, Field
from backend.app.db.models.enums import QuestionTypeEnum, DifficultyLevelEnum, BloomLevelEnum, QuestionStatusEnum

class MCQOptionsSchema(BaseModel):
    options: List[str] = Field(..., min_length=2)
    correct_option: str
    explanation: Optional[str] = None

class ValidationDataSchema(BaseModel):
    is_valid: bool = True
    alignment_score: float = 1.0
    difficulty_score: float = 1.0
    bloom_alignment_score: float = 1.0
    duplicate_score: float = 0.0
    completeness_score: float = 1.0
    warnings: List[str] = []
    similar_question_id: Optional[str] = None

class QuestionBase(BaseModel):
    subject_id: str
    unit_id: Optional[str] = None
    topic_id: Optional[str] = None
    learning_outcome_id: Optional[str] = None
    question_text: str = Field(..., min_length=1)
    question_type: QuestionTypeEnum
    marks: float = Field(2.0, gt=0)
    difficulty: DifficultyLevelEnum = DifficultyLevelEnum.MEDIUM
    bloom_level: BloomLevelEnum = BloomLevelEnum.UNDERSTAND
    expected_answer: Optional[str] = None
    options: Optional[Dict[str, Any]] = None
    keywords: Optional[List[str]] = None
    concepts: Optional[List[str]] = None
    source: Optional[str] = "MANUAL"
    status: QuestionStatusEnum = QuestionStatusEnum.ACTIVE

class QuestionCreate(QuestionBase):
    pass

class QuestionUpdate(BaseModel):
    subject_id: Optional[str] = None
    unit_id: Optional[str] = None
    topic_id: Optional[str] = None
    learning_outcome_id: Optional[str] = None
    question_text: Optional[str] = Field(None, min_length=1)
    question_type: Optional[QuestionTypeEnum] = None
    marks: Optional[float] = Field(None, gt=0)
    difficulty: Optional[DifficultyLevelEnum] = None
    bloom_level: Optional[BloomLevelEnum] = None
    expected_answer: Optional[str] = None
    options: Optional[Dict[str, Any]] = None
    keywords: Optional[List[str]] = None
    concepts: Optional[List[str]] = None
    status: Optional[QuestionStatusEnum] = None
    change_reason: Optional[str] = "Faculty edit"

class QuestionStatusUpdate(BaseModel):
    status: QuestionStatusEnum
    reason: Optional[str] = None

class QuestionVersionResponse(BaseModel):
    id: str
    question_id: str
    version_number: int
    question_text: str
    marks: float
    difficulty: DifficultyLevelEnum
    bloom_level: BloomLevelEnum
    expected_answer: Optional[str] = None
    keywords: Optional[Any] = None
    concepts: Optional[Any] = None
    changed_by: Optional[str] = None
    change_reason: Optional[str] = None
    created_at: Any

    model_config = ConfigDict(from_attributes=True)

class QuestionResponse(QuestionBase):
    id: str
    generation_run_id: Optional[str] = None
    version: int = 1
    validation_data: Optional[Dict[str, Any]] = None
    created_by: Optional[str] = None
    versions: List[QuestionVersionResponse] = []
    created_at: Any
    updated_at: Any

    model_config = ConfigDict(from_attributes=True)

class QuestionListResponse(BaseModel):
    total: int
    items: List[QuestionResponse]
    page: int
    page_size: int
